File: runtime/src/main/java/com/epam/drill/auto/test/agent/penetration/jmeter/JMeterPenetration.java
package com.epam.drill.auto.test.agent.penetration.jmeter;

import com.epam.drill.auto.test.agent.AgentClassTransformer;
import com.epam.drill.auto.test.agent.penetration.Strategy;
import javassist.CannotCompileException;
import javassist.CtClass;
import javassist.CtMethod;
import javassist.NotFoundException;

import java.io.IOException;


@SuppressWarnings("all")
public class JMeterPenetration extends Strategy {

    private String testNameSourceClass = "org.apache.jmeter.protocol.http.sampler.HTTPHC4Impl";

    @Override
    public String id() {
        return "jmeter";
    }

    public boolean permit(CtClass ctClass) {
        return ctClass.getName().equals(testNameSourceClass);
    }

    public byte[] instrument(CtClass ctClass) throws CannotCompileException, IOException, NotFoundException {
        CtMethod setupRequestMethod = ctClass.getMethod(
                "setupRequest",
                "(Ljava/net/URL;Lorg/apache/http/client/methods/HttpRequestBase;" +
                        "Lorg/apache/jmeter/protocol/http/sampler/HTTPSampleResult;)V"
        );
        setupRequestMethod.insertBefore(
                "String drillTestName = $3.getSampleLabel();\n" +
                        AgentClassTransformer.CLASS_NAME + ".memorizeTestName(drillTestName);\n"
        );
        return ctClass.toBytecode();
    }

}

File: src/runtimeMain/kotlin/com/epam/drill/test/agent/penetration/testing/jmeter/JMeterPenetration.kt
package com.epam.drill.test.agent.penetration.testing.jmeter

import com.epam.drill.test.agent.AgentClassTransformer
import com.epam.drill.test.agent.penetration.AbstractTestStrategy
import javassist.CtClass

class JMeterPenetration : AbstractTestStrategy() {
    private val testNameSourceClass = "org.apache.jmeter.protocol.http.sampler.HTTPHC4Impl"
    override val id: String = "jmeter"

    override fun permit(ctClass: CtClass): Boolean {
        return ctClass.name == testNameSourceClass
    }

    override fun instrument(ctClass: CtClass): ByteArray? {
        val setupRequestMethod = ctClass.getMethod(
            "setupRequest",
            "(Ljava/net/URL;Lorg/apache/http/client/methods/HttpRequestBase;" +
                    "Lorg/apache/jmeter/protocol/http/sampler/HTTPSampleResult;)V"
        )
        setupRequestMethod.insertBefore(
            """
                String drillTestName = $3.getSampleLabel();
                ${AgentClassTransformer.CLASS_NAME}.memorizeTestName(drillTestName);
                
                """.trimIndent()
        )
        return ctClass.toBytecode()
    }
}
